File: dijkstra.py
import heapq

def dijkstra(graph, start):
    dist = {node: float('inf') for node in graph}
    dist[start] = 0
    pq = [(0, start)]

    while pq:
        current_dist, node = heapq.heappop(pq)

        for neighbor, weight in graph[node]:
            distance = current_dist + weight

            if distance < dist[neighbor]:
                dist[neighbor] = distance
                heapq.heappush(pq, (distance, neighbor))

    return dist


n = int(input("Enter number of nodes: "))
graph = {}

for i in range(n):
    node = input(f"Enter node {i+1}: ")
    edges = int(input(f"Enter number of neighbors for {node}: "))

    neighbors = []
    for j in range(edges):
        neighbor = input("Enter neighbor node: ")
        weight = int(input("Enter weight: "))
        neighbors.append((neighbor, weight))

    graph[node] = neighbors


start = input("Enter start node: ")

result = dijkstra(graph, start)

print("\nShortest distances from", start)
for node in result:
    print(node, ":", result[node])

File: ai_practicals_cpp/dijkstra.cpp
#include <iostream>

using namespace std;

int main()
{
    int n;

    cout << "Enter number of vertices: ";
    cin >> n;

    int graph[10][10];

    cout << "Enter adjacency matrix:\n";

    for(int i=0; i<n; i++)
    {
        for(int j=0; j<n; j++)
        {
            cin >> graph[i][j];
        }
    }

    int visited[10] = {0};
    int distance[10];

    for(int i=0; i<n; i++)
    {
        distance[i] = 9999;
    }

    int start;

    cout << "Enter starting vertex: ";
    cin >> start;

    distance[start] = 0;

    for(int count=0; count<n-1; count++)
    {
        int min = 9999;
        int u;

        for(int i=0; i<n; i++)
        {
            if(!visited[i] && distance[i] < min)
            {
                min = distance[i];
                u = i;
            }
        }

        visited[u] = 1;

        for(int v=0; v<n; v++)
        {
            if(graph[u][v] &&
               !visited[v] &&
               distance[u] + graph[u][v] < distance[v])
            {
                distance[v] = distance[u] + graph[u][v];
            }
        }
    }

    cout << "\nShortest Distances:\n";

    for(int i=0; i<n; i++)
    {
        cout << start << " -> "
             << i << " = "
             << distance[i] << endl;
    }

    return 0;
}
